### PR TITLE
Fix for non-resvport auth and encryption in recent tpp changes

### PR DESCRIPTION
**Pull request by hirenvadalia**
#### Describe Bug or Feature

* With recent TPP changes non-resvport authentication and encryption is broken due to:
  * Double free of auth handshake data
  * Sender encrypts each chunk of pkt separately before sending, while receiver tries to decrypt whole pkt's encrypted data as it has got only two chunk - one encryption header chunk and another is encrypted data chunk
  * Invalid data pointer use in `*_pkt_handler_inner()` while encryption is enabled

#### Describe Your Change

* Fixed double free by removing unwanted `free(data_out)` in `tpp_handle_auth_handshake()`
* Fixed `tpp_encrypt_pkt()` to encrypt all chunks of pkt in one shot instead encrypting each chunk separately
* Fixed `*_pkt_handler_inner()` to use correct data pointer (aka `dhdr` instead `buf`) - as after decryption `dhdr` is pointing to decrypted data while `buf` is still pointing to original encrypted data

#### Link to Design Doc

* None

#### Attach Test and Valgrind Logs/Output

* In comment

### CODE REVIEW

**Review by prakashcv13 — APPROVED** [review]

**Comment by hirenvadalia** [comment]
[all-in-one.txt](https://github.com/openpbs/openpbs/files/5734065/all-in-one.txt)
[remote-mom.txt](https://github.com/openpbs/openpbs/files/5734066/remote-mom.txt)
[remote-mom-comm.txt](https://github.com/openpbs/openpbs/files/5734067/remote-mom-comm.txt)
[remote-mom-multi-comm-hop.txt](https://github.com/openpbs/openpbs/files/5734068/remote-mom-multi-comm-hop.txt)

Regression results:
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|5414|3839|0|0|0|0|3839|

File: src/lib/Libtpp/tpp_util.c [hunk 1]
```diff
@@ -2332,61 +2330,65 @@ tpp_encrypt_pkt(conn_auth_t *authdata, tpp_packet_t *pkt)
...
-		/* now replace the data with encrypted buffer */
-		free(chunk->data);
-		chunk->data = data_out;
-		chunk->len = len_out;
-		chunk->pos = chunk->data;
+	if (type == TPP_AUTH_CTX && data->for_encrypt == FOR_ENCRYPT)
+		return 0;
 
-		pkt->totlen += len_out;
+	buf = malloc(totlen);
```
Comment (subhasisb) [inline src/lib/Libtpp/tpp_util.c:2344]: oh but we will again malloc stuff. i think we should avoid this

**Review by subhasisb — CHANGES_REQUESTED** [review]
is there a better way rather than making one full packet from the chunks and then encrypting it? I think we should avoid the additional malloc if possible

File: src/lib/Libtpp/tpp_util.c (hunk 1 above)
Comment (hirenvadalia) [reply src/lib/Libtpp/tpp_util.c:2344]: yes agree but I couldn't find any better way unless we somehow make chunks on receiver side instead full buffer... will chat offline...

**Review by subhasisb — APPROVED** [review]
After speaking to Hiren, I agree to his changes. Approving

**Pull request merged**